### PR TITLE
Double quote escape for bash script ENV_VAR part

### LINKED ISSUE #4836: Consider using double quotes for the values in the `metadata.options.environment_variables` dict
If one specifies something like:
`"environment_variables":{"MY_FILE":"$HOME/path/my_file"},`
in the `metadata.options`, the submission script would include:
```
# ENVIRONMENT VARIABLES BEGIN ###
export MY_FILE='$HOME/path/my_file'
# ENVIRONMENT VARIABLES  END  ###
```
This call is ineffective since the  `$` would be considered a normal character.
The desired behavior would be to return double quotes:
 ```
# ENVIRONMENT VARIABLES BEGIN ###
export MY_FILE="$HOME/path/my_file"
# ENVIRONMENT VARIABLES  END  ###
```
since within double quotes the special character `$` is recognized.
For reference, here an extract of the bash manual:
```
    3.1.2.2 Single Quotes

    Enclosing characters in single quotes (') preserves the literal value of
    each character within the quotes.
    A single quote may not occur between single quotes, even when preceded by
    a backslash.

    3.1.2.3 Double Quotes

    Enclosing characters in double quotes (") preserves the literal value of all characters
    within the quotes, with the exception of $, `, \, and, when history expansion is enabled, !.
    The characters $ and ` retain their special meaning within double quotes (see Shell Expansions).
    The backslash retains its special meaning only when followed by one of the following characters:
    $, `, ", \, or newline. Within double quotes, backslashes that are followed by one of these
    characters are removed. Backslashes preceding characters without a special meaning are left
    unmodified. A double quote may be quoted within double quotes by preceding it with a backslash.
    If enabled, history expansion will be performed unless an ! appearing in double quotes is escaped
    using a backslash. The backslash preceding the ! is not removed.
    The special parameters * and @ have special meaning when in double quotes
    (see Shell Parameter Expansion).
```

The solution requires some improvements in the method `escape_for_bash`.
https://github.com/aiidateam/aiida-core/blob/75310e0bfbf8de02302de76f3e452b38e4164f33/aiida/common/escaping.py#L15

### PR DESCRIPTION
**Pull request by unkcpz**
This PR is separated from https://github.com/aiidateam/aiida-core/pull/5280 and containerized code PR for easy review. 

Fixes #4836

### CODE REVIEW

File: tests/common/test_escaping.py [hunk 1]
```diff
@@ -24,3 +28,23 @@ def test_escape_for_bash():
...
+        ['string', '"string"'],
+        ['string with space', '"string with space"'],
+        ['string with a " double quote', '''"string with a "'"'" double quote"'''],
+        [1, '"1"'],
+        [2.0, '"2.0"'],
+        ['$PWD', '"$PWD"'],
+    )
+
+    for string_input, string_escaped in tests:
+        assert escape_for_bash(string_input, use_double_quotes=True) == string_escaped
```
Comment (sphuber) [inline tests/common/test_escaping.py:50]: This is the perfect candidate to use `pytest.mark.parametrize. I would suggest replacing the original test and the one you added with the following single test:
```suggestion
import pytest

from aiida.common.escaping import escape_for_bash


@pytest.mark.parametrize(('to_escape, expected_single_quotes, expected_double_quotes'), (
    (None, ''),
    ('string', "'string'", '"string"'),
    ('string with space', "'string with space'", '"string with space"'),
    ("string with ' single quote", """'string with '"'"' single quote'""", '''"string with "'"'" double quote"'''),
    (1, "'1'", '"1"'),
    (2.0, "'2.0'", '"2.0"'),
    ('$PWD', "'$PWD'", '"$PWD"'),
))
def test_escape_for_bash(to_escape, expected_single_quotes, expected_double_quotes):
    """Tests various inputs for `aiida.common.escaping.escape_for_bash`."""
    assert escape_for_bash(to_escape, use_double_quotes=False) == expected_single_quotes
    assert escape_for_bash(to_escape, use_double_quotes=True) == expected_double_quotes
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @unkcpz. Looks good. Just a suggestion to clean up the test. And I personally am still not convinced that `envvar_double_quotes` is the best name. But if others think this is fine, then I won't block it.

File: tests/common/test_escaping.py (hunk 1 above)
Comment (unkcpz) [reply tests/common/test_escaping.py]: Thanks @sphuber! Absolutely it is beautiful. One line to rule all test cases 😄 

I also change the name to `environment_variables_double_quotes` as you suggest.

File: aiida/schedulers/datastructures.py [hunk 2]
```diff
@@ -248,6 +248,9 @@ class JobTemplate(DefaultFieldsAttributeDict):  # pylint: disable=too-many-insta
       * ``rerunnable``: if the job is rerunnable (boolean)
       * ``job_environment``: a dictionary with environment variables to set
         before the execution of the code.
+      * ``environment_variables_double_quotes``: if set to True, use double quotes
+        instead of single quotes to escape the environment variables specified
+        in ``environment_variables``.
```
[the remark this answers is not in the document]
Comment (sphuber) [reply aiida/schedulers/datastructures.py:253]: @unkcpz just noticed this last mistake in the docstring. Sorry for having missed it. If you correct this, I will merge the PR. Thanks a lot

[the remark this answers is not in the document]
Comment (unkcpz) [reply aiida/schedulers/datastructures.py:253]: @sphuber Thanks. Updated.

**Review by sphuber — APPROVED** [review]

**Pull request merged**